### PR TITLE
feat: Read configuration from files, Secrets, and env variables

### PR DESCRIPTION
**Pull request by jkayani**
### Description

Previously we read configuration for Vault from env variables. Now we can read from files on disk, and from Kubernetes Secrets directly. 

Accomplished by adding 2 new CLI flags:

- `-c` to specify the path to a config file
- `-s` to specify the name of a secret in `argocd` namespace of the ArgoCD cluster

A config file can be in any format Viper accepts (JSON, YAML, etc). The keys must be the ones shown in this YAML example:
```
VaultType: vault
Address: http://a-vault/
PathPrefix: my-app
AuthType: github
GithubToken: token
RoleID: blah
SecretID: big-blah
IBMCloudAPIKey: biggest-blah
```

A Secret just has to have the above keys:
```
apiVersion: v1
data:
  Address: aHR0cHM6Ly92YXVsdC5ib2Iv
  AuthType: Z2l0aHVi
  GithubToken: Zm9v
  PathPrefix: YmxhaAo=
  VaultType: dmF1bHQ=
kind: Secret
metadata:
  name: bob
  namespace: argocd
type: Opaque
```

Set env variables take precedence over config files and k8s Secrets via Viper. The env variable names we accept should be the same exact ones accepted before. Note that reading k8s Secrets from the plugin requires the `argocd-repo-server` to have a service account token mounted in the standard location `/var/run/secrets/kubernetes.io/`, which by default it doesn't do for some reason.

Also added error messages when a required configuration parameter is left empty; i.e, `GithubToken` for `github` authentication into a `vault` Vault. 

**Fixes:** #32 

### Checklist
Please make sure that your PR fulfills the following requirements: 
- [ ] Reviewed the guidelines for contributing to this repository
- [ ] The commit message follows the [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] Tests for the changes have been
- [ ] Docs have been added / updated

### Type of Change

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] New tests
- [ ] Build/CI related changes
- [ ] Documentation content changes
- [ ] Other (please describe)

### CODE REVIEW

File: pkg/vault/config.go [hunk 1]
```diff
@@ -16,18 +16,31 @@ type Config struct {
 	Type       VaultType
 }
 
+func setupViper() {
+	viper.BindEnv("VaultType", "AVP_TYPE")
```
Comment (werne2j) [inline pkg/vault/config.go:20]: We should be able to use the `SetEnvPrefix` to simplify this.
https://github.com/spf13/viper#working-with-environment-variables

File: pkg/client/kubernetes.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	"k8s.io/client-go/rest"
+	k8yaml "sigs.k8s.io/yaml"
+)
+
+type Client struct {
+	client *kubernetes.Clientset
+}
+
+func NewClient() (*Client, error) {
+	config, err := rest.InClusterConfig()
```
Comment (werne2j) [inline pkg/client/kubernetes.go:15]: Can we have a nice error in the situation where someone tries to do this while not in kubernetes

File: cmd/generate.go [hunk 3]
```diff
@@ -71,5 +96,7 @@ func NewGenerateCommand() *cobra.Command {
 		},
 	}
 
+	command.Flags().StringVarP(&configPath, "config-path", "c", "", "path to a configuration file (YAML, JSON, envfile) to use")
+	command.Flags().StringVarP(&secretName, "secret-name", "s", "", "name of a Kubernetes Secret containing Vault configuration data in the argocd namespace of your ArgoCD host")
```
Comment (werne2j) [inline cmd/generate.go:100]: Can you add something to make it very clear that `secret-name` will only work if in ArgoCD?

**Review by werne2j — COMMENTED** [review]
Overall looks pretty good. Just a few comments.

File: pkg/vault/config.go (hunk 1 above)
Comment (jkayani) [reply pkg/vault/config.go:20]: I looked at that but:

> If the ENV variable name is not provided, then Viper will automatically assume that the ENV variable matches the following format: prefix + "_" + the key name in ALL CAPS. When you explicitly provide the ENV variable name (the second parameter), it does not automatically add the prefix.

We'd have to either rename the Viper keys from `RoleID` to `Role_ID` to match the env var name `AVP_ROLE_ID` or rename the env variables. We can go either way I just didn't do them.

File: pkg/client/kubernetes.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
+package client
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	k8yaml "sigs.k8s.io/yaml"
+)
+
+type Client struct {
```
Comment (jkayani) [inline pkg/client/kubernetes.go:12]: Is there a better name for this? Maybe package `localk8s`?

File: pkg/vault/config.go (hunk 1 above)
Comment (jkayani) [reply pkg/vault/config.go:20]: Depending on how we answer this, we may or may not need to add docs that say what the accepted keys in a config file /Secret are

File: pkg/client/kubernetes.go (hunk 2 above)
Comment (jkayani) [reply pkg/client/kubernetes.go:21]: Added

File: cmd/generate.go (hunk 3 above)
Comment (werne2j) [reply cmd/generate.go:100]: I was just looking for something like `(Only available when used in ArgoCD)` to make it very obvious. Someone might read this `name of a Kubernetes Secret containing Vault configuration data in the argocd namespace of your ArgoCD host` and not really understand.

File: pkg/vault/config.go (hunk 1 above)
Comment (werne2j) [reply pkg/vault/config.go:20]: > rename the Viper keys from RoleID to Role_ID
I am fine with this

**Comment by jkayani** [comment]
Seems this code is failing the `go vet` check with:

```
# github.com/hashicorp/vault/vault/quotas
../../../go/pkg/mod/github.com/hashicorp/vault@v1.6.1/vault/quotas/quotas_rate_limit.go:280:33: assignment mismatch: 4 variables but rlq.store.Take returns 5 values
../../../go/pkg/mod/github.com/hashicorp/vault@v1.6.1/vault/quotas/quotas_rate_limit.go:280:50: not enough arguments in call to rlq.store.Take
	have (string)
	want (context.Context, string)
../../../go/pkg/mod/github.com/hashicorp/vault@v1.6.1/vault/quotas/quotas_rate_limit.go:305:25: not enough arguments in call to rlq.store.Close
	have ()
	want (context.Context)
```

I'm not really sure what to do here. I tried setting the Vault package to 1.6.1 since it was 1.6.0 before and one of my previous commits downgraded it to 1.4.something to fix this.

File: pkg/client/kubernetes.go (hunk 4 above)
Comment (werne2j) [reply pkg/client/kubernetes.go:15]: Can be renamed, not a fan of `LocalK8s` can be `KubernetesClient` or something

**Comment by werne2j** [comment]
Maybe we should not run vet on the dependencies

File: README.md [hunk 5]
```diff
@@ -42,17 +42,58 @@ data:
...
+The plugin requires some configuration to connect to Vault. The parameters are:
 
-| Name            | Description                | Supported            |
+| Name            | Description                | Notes                |
 | --------------- | -------------------------- | -------------------- |
-| AVP_VAULT_ADDR  | Address of your Vault      | N/A                  |
-| AVP_PATH_PREFIX | Prefix of the vault path to look for the secrets | N/A       |
-| AVP_TYPE        | The type of Vault backend  | `vault` and `secretmanager` |
-| AVP_AUTH_TYPE   | The type of authentication | vault: `approle, github`   secretmanager: `iam` |
+| VAULT_ADDR     | Address of your Vault      | N/A                  |
```
Comment (werne2j) [inline README.md:50]: These still have to have the prefix, right?

Comment (jkayani) [reply README.md:50]: Yeah only when specified as env variables

Comment (werne2j) [reply README.md:50]: Oh so its not needed when a config or secret, got it

**Review by werne2j — APPROVED** [review]
Looks good

**Pull request merged**